src/Save/Save.man
.\" t
.\" @(#)Save.1	12/10/94
.TH Save 1.8 "Feb 7 2000" Save
.UC
.SH NAME
\fBSave\fP \- the AfterStep desktop-layout saving module

.SH SYNOPSIS
\fBSave\fP [\-f \fIconfig_file\fP] [\-v | \--version] [\-h | \--help]
.sp
Although \fBSave\fP can be run from the command line, it is normally 
spawned by afterstep from within an afterstep menu file by 'Save':
.nf
.sp
% cat "~/GNUstep/Library/AfterStep/start/Modules/Save"
	Module "Save" Save
	MiniPixmap "mini-as.xpm"
.fi

.SH DESCRIPTION
When called, this module will attempt to save your current desktop
layout into a file called .xinitrc.
Usually, you will have to edit it to get a useable configuration, so be sure to
keep a backup of your old .xinitrc.
 
Your applications must supply certain hints to the X window system.
Emacs, for example, does not, so Save can't get any information from it.  
 
Also, Save assumes that certain command line options are
globally accepted by applications, which may not be the case.

AND DON'T FORGET AFTERSTEP ALREADY SAVES DESKTOP INTO
~/GNUstep/Library/AfterStep/.workspace_state file, called from autoexec !

Using both Save and AfterStep default saving will save twice the same window,
therefore open it twice on next startup !

.SH OPTIONS
.IP "\-f \fIconfig_file\fP"
Use \fIconfig_file\fP instead of the default configuration file.

.IP "\-h | \--help"
Print a usage message and exit successfully.

.IP "\-v | \--version"
Print version information and exit successfully.

.SH INITIALIZATION
During initialization, \fISave\fP will eventually search a configuration 
file which describes the time-outs and actions to take.  This file 
will be the one specified by the \fI\-f\fP option (if given).  Else, the 
~/GNUstep/Library/AfterStep/save file will be used if it exists.  Else, 
/usr/local/share/afterstep/save will be used.

.SH INVOCATION
Save can be invoked by inserting the line 'Save' in
"~/GNUstep/Library/AfterStep/start/Modules/Save desktop" file.
It can be bound to a menu or mouse button or keystroke in feel file to invoke
it later.
AfterStep will search directory specified in the ModulePath configuration
option to attempt to locate Save.

.SH CONFIGURATION OPTIONS
Sorry, but there is none at the moment.
Moreover, which options could be useful ?

.SH AUTHORS
Per Persson <pp@solace.mh.se>
.nf
Robert Nation
.nf
David Mihm <dmihm@swbell.net> [Man page]

.SH SEE ALSO
.BR afterstep (1)


ChangeLog

2003-09-25  sasha

	* generating ChangeLog

2003-09-24  sasha

	* Completed look.Glass ; renamed PointerFore, PointerBack to
	  CursorFore, CursorBack to match old Cursor option; Moved Cursor
	  definitions from feel into look Completed Colorscheme's support
	  for Cursor's color

2003-09-23  sasha

	* Added look options for mouse pointer coloring: PointerFore and
	  PointerBack; Repaired code to update window's decorations when
	  stickiness flag is toggled;

2003-09-22  sasha

	* Added cursors to default feel file; enable cursor switching while
	  hovering over different elements of the interface; Added code to
	  change color of the cursor according to what colorscheme is set
	  to - may need to tweak a bit , or make colors configurable
	* Fixed locale parsing from LANG variable; Fixed module startup to
	  pass on current locae specified on command line

2003-09-19  sasha

	* added debugging code for cursor loading

2003-09-16  sasha

	* fixed bug with windows teleporting back to original position, if
	  moved/resized too fast

2003-09-15  sasha

	* updated configure scripts
	* Fixed configure.in to properly enable i18n in libAfterImage as
	  requested. Internationalized text should be displayed properly
	  now, even if its not in UTF8; Update src/test to work with new
	  libs and added test code for i18n testing
	* Fixed unquoted string in i18N code reported by Alex Mezin

2003-09-14  sasha

	* Added checks for locale and proper initialization of
	  libAfterImage t3ext drawing in accordance to its value; Fixed
	  problem with EdgeScroll being mangled when just look changes;
	  Fixed recently introduced bug causing wierd corners in framned
	  windows upon the look change
	* Fixed typoe for msg type in SendString being declared as some
	  wierd type of unsigned long long  - possibly causing wierdness on
	  64 bit machines

2003-09-13  sasha

	* implemented feedback functionality via sending the e-mail to
	  sasha :)

2003-09-12  sasha

	* added configure option to send e-mail back to developer on
	  installation; Fixed canvas border width handling while resizing
	  and in Pager;

2003-09-11  sasha

	* fixed handling of WM_PROTOCOL changes where WM_COMMAND would get
	  lost and session management would stop working; Found bug with
	  huge impact - all the geometry calculations were off as we were
	  forgetting border width of the frame window; Fixed handling of
	  button presses on border of the frame

2003-09-10  sasha

	* fixed NeXTish colorscheme somewhat
	* Fixed workspace state save/restore to allow terms to exec command
	  using -e, and to have argv's with spaces in them; Fixed frame
	  corner sizing to make it same width as is tbar size;

2003-09-09  sasha

	* More fixing Pager's geometry; Added -i cmd line parameter to
	  ascompose to allow for 'include' files (such as colorschemes,
	  etc.); Updated configure scripts to properly detect headers when
	  its run from libAfterBase; Updated version to beta2; Updated TEAM
	  file

2003-09-08  sasha

	* Fixed Pager to properly account for label size while sizing
	  min-desktop windows

2003-09-07  sasha

	* somewhat debugged Pager's geometry handling - need to add some
	  code to properly recalculate Pager's geometry if its soecified as
	  ratio of the main desktop vs. exact geometry

2003-09-03  sasha

	* added new AfterStep log icon; fixed wharf config file

2003-09-02  sasha

	* added new xml tag : <color> to define symbolic color names
	* changed handling of tiled images to clip it properly when
	  Right/Left align is used

2003-08-30  sasha

	* debugged workspace state handling code, and improved Wait loop

2003-08-29  sasha

	* mostly implemented new workspace state saving and restoring

2003-08-28  sasha

	* fixed workspace status generator to not save xrm params, as noone
	  seems to support thoise
	* added caching of pixmaps for unfocused bas in window decorations
	  to speed up fast focus switching
	* 64-bit fixes: CARD32 instead of Window type
	* 64-bit fixes: properties and strlen warnings
	* 64-bit fixes: converted intermodule communications to always use
	  32bit values
	* more 64 bit fixes

2003-08-27  sasha

	* 64-bit compilation fixes
	* Fixed slowness in intermodule communications manifesting itself
	  when desktops are switched
	* fixed handling of non-default MyBackgrounds, and proper adding of
	  DeskConfigs while switching desks
	* synching
	*  fixed segfaulting in Pager on look switches or after window
	  restacking
	* Fixed handling of hints flags - it was doing nothing before if
	  flags == 0 - was screwing MOtif Hints like totatlly; Fixed shaped
	  window handling in WinList; Added check to not do too much
	  redrawing when WinLIst is receiving window list - may want to do
	  the same in Pager

2003-08-26  sasha

	*  added parameters to import funcions, to allow for sized
	  image/preview stuff; Removed compression buffer from ASImage
	  structure to save up some memory; Hopefully fixed hints handling
	  on 64 bit hw

2003-08-24  sasha

	* fixed event handling when windows is getting mapped - no more
	  lost events due to not selected mask ; Fixed initial status when
	  mouse pointer is out of the screen; Fixed ICCCM size hints
	  handling to have BaseSize and MinSize interchange each other when
	  one is ommited - fixes GIMPs initial placement

2003-08-23  sasha

	* added new 3 icons to complete the set of default icons: switched
	  AfterStep to use ~/.afterstep instead of ridiculous ~/GNUstel/L/A
	  stuff

2003-08-22  sasha

	* Fixed handling of shaped windows - shape is not erroneously
	  cleared, and client's shape is subtracted from decoration's shape
	* fixed autodetection of zipped xpms and added 6 more icons to
	  default iconset

2003-08-21  sasha

	* more improvements to asimbrowser

2003-08-20  sasha

	* fixed makefile for ascp, when fltk is not found; added better
	  Style for BitchX to database; remnoved scrollbar from BitchX in
	  Wharf

src/Save/Save.c
/*
 * Copyright (c) 1997 Guylhem Aznar <guylhem@oeil.qc.ca>
 * Copyright (c) 1994 Robert Nation
 * Copyright (c) 1994 Per Persson <pp@solace.mh.se>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#define TRUE 1
#define FALSE

#include "../../configure.h"

#include <stdio.h>
#include <signal.h>
#include <fcntl.h>
#include <string.h>
#include <sys/wait.h>
#include <sys/time.h>
#include <unistd.h>
#include <ctype.h>
#include <stdlib.h>
#include <X11/Xlib.h>
#include <X11/Xutil.h>
#include <X11/Xproto.h>
#include <X11/Xatom.h>
#include <X11/Intrinsic.h>

#include "../../include/aftersteplib.h"
#include "../../include/module.h"

#include "Save.h"

char *MyName;
int fd_width;
int fd[2];

struct list *list_root = NULL;

Display *dpy;			/* which display are we talking to */
int ScreenWidth, ScreenHeight;
int screen;

long Vx, Vy;

void
version (void)
{
  printf ("%s version %s\n", MyName, VERSION);
  exit (0);
}

void
usage (void)
{
  printf ("Usage:\n"
	  "%s [--version] [--help]\n", MyName);
  exit (0);
}

/***********************************************************************
 *
 *  Procedure:
 *	main - start of module
 *
 ***********************************************************************/
int
main (int argc, char **argv)
{
  char *temp;
  int i;
  char *global_config_file = NULL;

  /* Save our program name - for error messages */
  temp = strrchr (argv[0], '/');
  MyName = temp ? temp + 1 : argv[0];

  for (i = 1; i < argc && *argv[i] == '-'; i++)
    {
      if (!strcmp (argv[i], "-h") || !strcmp (argv[i], "--help"))
	usage ();
      else if (!strcmp (argv[i], "-v") || !strcmp (argv[i], "--version"))
	version ();
      else if (!strcmp (argv[i], "-w") || !strcmp (argv[i], "--window"))
	i++;
      else if (!strcmp (argv[i], "-c") || !strcmp (argv[i], "--context"))
	i++;
      else if (!strcmp (argv[i], "-f") && i + 1 < argc)
	global_config_file = argv[++i];
    }

  /* Dead pipe == dead AfterStep */
  signal (SIGPIPE, DeadPipe);

  if ((dpy = XOpenDisplay ("")) == NULL)
    {
      fprintf (stderr, "%s: couldn't open display %s\n",
	       MyName, XDisplayName (""));
      exit (1);
    }
  screen = DefaultScreen (dpy);

  /* connect to AfterStep */
  temp = module_get_socket_property (RootWindow (dpy, screen));
  fd[0] = fd[1] = module_connect (temp);
  XFree (temp);
  if (fd[0] < 0)
    {
      fprintf (stderr, "%s: unable to establish connection to AfterStep\n", MyName);
      exit (1);
    }
  temp = safemalloc (9 + strlen (MyName) + 1);
  sprintf (temp, "SET_NAME %s", MyName);
  SendInfo (fd, temp, None);
  free (temp);

  ScreenHeight = DisplayHeight (dpy, screen);
  ScreenWidth = DisplayWidth (dpy, screen);

  fd_width = GetFdWidth ();

  /* Create a list of all windows */
  /* Request a list of all windows,
   * wait for ConfigureWindow packets */
  SendInfo (fd, "Send_WindowList", 0);

  Loop (fd);

  return 0;
}


/***********************************************************************
 *
 *  Procedure:
 *	Loop - wait for data to process
 *
 ***********************************************************************/
void
Loop (int *fd)
{
  unsigned long header[3], *body;
  int count = 0;

  while (1)
    {
      /* read a packet */
      if ((count = ReadASPacket (fd[1], header, &body)) > 0)
	{
	  /* dispense with the new packet */
	  process_message (header[1], body);
	  free (body);
	}
    }
}


/***********************************************************************
 *
 *  Procedure:
 *	Process message - examines packet types, and takes appropriate action
 *
 ***********************************************************************/
void
process_message (unsigned long type, unsigned long *body)
{
  switch (type)
    {
    case M_CONFIGURE_WINDOW:
      if (!find_window (body[0]))
	add_window (body[0], body);
      break;
    case M_NEW_PAGE:
      list_new_page (body);
      break;
    case M_END_WINDOWLIST:
      do_save ();
      break;
    default:
      break;
    }
}





/***********************************************************************
 *
 *  Procedure:
 *	find_window - find a window in the current window list 
 *
 ***********************************************************************/
struct list *
find_window (unsigned long id)
{
  struct list *l;

  if (list_root == NULL)
    return NULL;

  for (l = list_root; l != NULL; l = l->next)
    {
      if (l->id == id)
	return l;
    }
  return NULL;
}



/***********************************************************************
 *
 *  Procedure:
 *	add_window - add a new window in the current window list 
 *
 ***********************************************************************/
void
add_window (unsigned long new_win, unsigned long *body)
{
  struct list *t;

  if (new_win == 0)
    return;

  t = (struct list *) safemalloc (sizeof (struct list));
  t->id = new_win;
  t->next = list_root;
  t->frame_height = (int) body[6];
  t->frame_width = (int) body[5];
  t->base_width = (int) body[11];
  t->base_height = (int) body[12];
  t->width_inc = (int) body[13];
  t->height_inc = (int) body[14];
  t->frame_x = (int) body[3];
  t->frame_y = (int) body[4];;
  t->title_height = (int) body[9];;
  t->boundary_width = (int) body[10];
  t->flags = (unsigned long) body[8];
  t->gravity = body[21];
  list_root = t;
}



/***********************************************************************
 *
 *  Procedure:
 *	list_new_page - capture new-page info
 *
 ***********************************************************************/
void
list_new_page (unsigned long *body)
{
  Vx = (long) body[0];
  Vy = (long) body[1];
}
/***********************************************************************
 *
 *  Procedure:
 *	SIGPIPE handler - SIGPIPE means AfterStep is dying
 *
 ***********************************************************************/
void
DeadPipe (int nonsense)
{
  exit (0);
}


/***********************************************************************
 *
 *  Procedure:
 *	writes a command line argument to file "out"
 *      checks for qoutes and stuff
 *
 ***********************************************************************/
void
write_string (FILE * out, char *line)
{
  int len, space = 0, qoute = 0, i;

  len = strlen (line);

  for (i = 0; i < len; i++)
    {
      if (isspace (line[i]))
	space = 1;
      if (line[i] == '\"')
	qoute = 1;
    }
  if (space == 1)
    fprintf (out, "\"");
  if (qoute == 0)
    fprintf (out, "%s", line);
  else
    {
      for (i = 0; i < len; i++)
	{
	  if (line[i] == '\"')
	    fprintf (out, "\\\"");
	  else
	    fprintf (out, "%c", line[i]);
	}
    }
  if (space == 1)
    fprintf (out, "\"");
  fprintf (out, " ");

}



/***********************************************************************
 *
 *  Procedure:
 *	checks to see if we are supposed to take some action now,
 *      finds time for next action to be performed.
 *
 ***********************************************************************/
void
do_save (void)
{
  struct list *t;
  char tname[200], loc[30];
  FILE *out;
  char **command_list;
  int dwidth, dheight, xtermline = 0;
  int x1, x2, y1, y2, i, command_count;
  long tVx, tVy;

  sprintf (tname, "%s/.xinitrc", getenv ("HOME"));
  out = fopen (tname, "w+");
  for (t = list_root; t != NULL; t = t->next)
    {
      tname[0] = 0;

      x1 = t->frame_x;
      x2 = ScreenWidth - x1 - t->frame_width - 2;
      if (x2 < 0)
	x2 = 0;
      y1 = t->frame_y;
      y2 = ScreenHeight - y1 - t->frame_height - 2;
      if (y2 < 0)
	y2 = 0;
      dheight = t->frame_height - t->title_height - 2 * t->boundary_width;
      dwidth = t->frame_width - 2 * t->boundary_width;
      dwidth -= t->base_width;
      dheight -= t->base_height;
      dwidth /= t->width_inc;
      dheight /= t->height_inc;

      if (t->flags & STICKY)
	{
	  tVx = 0;
	  tVy = 0;
	}
      else
	{
	  tVx = Vx;
	  tVy = Vy;
	}
      sprintf (tname, "%dx%d", dwidth, dheight);
      if ((t->gravity == EastGravity) ||
	  (t->gravity == NorthEastGravity) ||
	  (t->gravity == SouthEastGravity))
	sprintf (loc, "-%d", x2);
      else
	sprintf (loc, "+%d", x1 + (int) tVx);
      strcat (tname, loc);

      if ((t->gravity == SouthGravity) ||
	  (t->gravity == SouthEastGravity) ||
	  (t->gravity == SouthWestGravity))
	sprintf (loc, "-%d", y2);
      else
	sprintf (loc, "+%d", y1 + (int) tVy);
      strcat (tname, loc);

      if (XGetCommand (dpy, t->id, &command_list, &command_count))
	{
	  for (i = 0; i < command_count; i++)
	    {
	      if (strncmp ("-geo", command_list[i], 4) == 0)
		{
		  i++;
		  continue;
		}
	      if (strncmp ("-ic", command_list[i], 3) == 0)
		continue;
	      if (strncmp ("-display", command_list[i], 8) == 0)
		{
		  i++;
		  continue;
		}
	      write_string (out, command_list[i]);
	      if (strstr (command_list[i], "xterm"))
		{
		  fprintf (out, "-geometry %s ", tname);
		  if (t->flags & ICONIFIED)
		    fprintf (out, "-ic ");
		  xtermline = 1;
		}
	    }
	  if (command_count > 0)
	    {
	      if (xtermline == 0)
		{
		  if (t->flags & ICONIFIED)
		    fprintf (out, "-ic ");
		  fprintf (out, "-geometry %s &\n", tname);
		}
	      else
		{
		  fprintf (out, "&\n");
		}
	    }
	  XFreeStringList (command_list);
	  xtermline = 0;
	}
    }
  fprintf (out, "\nexec afterstep\n");
  fclose (out);
  exit (0);
}
